---
title: "Data import, Seurat object creation and overview"
authors: "Rebecca Jaszczak, Karla Lindquist, Angelo Pelonero"
date: "3/2/2020"
output: html_notebook
--- 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

For this tutorial, we will be analyzing a dataset of Peripheral Blood Mononuclear Cells (PBMC) freely available from 10X Genomics. There are 2,700 single cells that were sequenced on the Illumina NextSeq 500. The raw data can be found [here](https://s3-us-west-2.amazonaws.com/10x.files/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz).

Other raw data (for your future explorations) are available [here](https://support.10xgenomics.com/single-cell-gene-expression/datasets).

We start by reading in the data. The `Read10X()` function reads in the output of the cellranger pipeline from 10X, returning a unique molecular identified (UMI) count matrix. The values in this matrix represent the number of molecules for each feature (i.e. gene; row) that are detected in each cell (column).

For more information about the cellranger pipeline, visit the [CellRanger webpage](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger).

We next use the count matrix to create a `Seurat` object. The object serves as a container that contains both data (like the count matrix) and analysis (like PCA, or clustering results) for a single-cell dataset. For a technical discussion of the Seurat object structure, check out their [GitHub Wiki](https://github.com/satijalab/seurat/wiki). For example, the count matrix is stored in pbmc[["RNA"]]@counts.

```{r}
library(dplyr)
library(Seurat)

# Load the PBMC dataset
pbmc.data <- Read10X(data.dir = "./data/filtered_gene_bc_matrices/hg19/")
# Initialize the Seurat object with the raw (non-normalized data).
pbmc <- CreateSeuratObject(counts = pbmc.data, project = "pbmc3k", min.cells = 3, min.features = 200)
pbmc
```
Now that the data is loaded into R, what does this data in a count matrix look like?

```{r}
# Lets examine a few genes in the first thirty cells
pbmc.data[c("CD3D", "TCL1A", "MS4A1"), 1:30]
```
The `.` values in the matrix represent 0s (no molecules detected). Since most values in an scRNA-seq matrix are 0, Seurat uses a sparse-matrix representation whenever possible. This results in significant memory and speed savings for Drop-seq/inDrop/10x data. This is especially important was datasets venture into millions of cells.

Let's take a quick look at what this means in practical terms:
```{r}
dense.size <- object.size(as.matrix(pbmc.data))
dense.size
```
~710 million bytes = 710 megabytes

```{r}
sparse.size <- object.size(pbmc.data)
sparse.size
```
~30 million bytes = 30 megabytes

```{r}
dense.size/sparse.size
```
The dense matrix is about 24 times larger than the sparse matrix format - thanks Seurat! This will become more relevant as we conduct complex analysis and our Seurat object (built off of the sparse matrix) has complex, high memory calculations added to it.